Opal Ice Maker Production/Progress Feature

jamesh48 opened the pull request:
## :recycle: Current situation

The initial Opal Ice Maker PR included basic support for controlling the power state and the nightlight state via homekit.

## :bulb: Proposed solution

The changes in this PR introduces a new Ice Production Progress feature.
The Opal Ice Maker exposes an OIM_PRODUCTION value that roughly translates to the percentage of ice production on a scale from 0 - 100. As the ice maker produces ice, this value raises sequentially. 

With this value I was able to introduce two new features:

1.  In homekit, a read only progress bar that updates automatically as ice is produced. 

I struggled a bit with finding the best representation of "progress" and settled on a Fan Service with the RotationSpeed Characteristic, admittedly it looks like something the user can set, but its actually readonly.

![IMG_5449](https://github.com/user-attachments/assets/095caf5b-43a7-4175-9d21-7faa15d52368)

![IMG_5450](https://github.com/user-attachments/assets/c8543c31-daef-4986-aa69-978c0d077746)

2. The ability for the user to set a desired ice production limit in the plugin settings, such that the ice maker shuts off automatically at the given limit rather than filling all the way to 100%. 

Tying the two features together - the progress bar displays relative to the users preferred production limit. So for example if the users preferred production limit is 20, then the progress bar dynamically increments in steps of 5, such that if the current production value is at 10, then the progress bar shows 50%.

The default production limit is 100.

In the above photos, I have my production limit set to 50, and for demonstration purposes hardcoded 30 as the production value. So the progress bar shows at 60%.

On the plugin side, I had to make updates to how the refreshRate is handled, to expose the homekit updateRate, and to find a clever way to pass metadata from the service to the plugin such that the user could set their desired production limit.

* Update Rate now exposed, minimum value of 30, allowing the user to set this lower than the default facilitates a more "live" presentation of the progress bar in the home app. The default seemed to be somewhere between 2-5 minutes and was painful. Ideally the deviceRefreshRate and updateRate are the same, so that when homebridge polls the ice machine that it in turn automatically updates home app.

<img width="621" alt="Screenshot 2025-02-26 at 10 07 26 AM" src="https://github.com/user-attachments/assets/e0d624b1-05c3-4f24-8e6f-656c07ac33ba" />

* I defined a "ProductData" Characteristic on the AccessoryInformation Service that is passed to the plugin index.html file. Its a bit hacky as the only way I could figure out how to make it compliant with HomeKit was to pass the value as a query string, like this `l=Production_Limit&i=OPL&t=number&d=100`, and parse it in the javascript into a JSON object. `l` is label, `i` is indicator (or key) for the option key value pair, `t` is type and `d` is default as in default value. It seems that there was a constraint on the length of this value so I had to use letters.

<img width="757" alt="Screenshot 2025-02-26 at 10 10 09 AM" src="https://github.com/user-attachments/assets/bd0d0234-b9f2-4c58-8cde-62f6093dd547" />

## :gear: Release Notes

Opal Ice Machine Production Progress Feature

## :heavy_plus_sign: Additional Information

https://developers.homebridge.io/#/

Because most users will probably be inclined to set their production limit and then forget it as it runs on a local server persistently, I did consider giving the user the ability to dynamically _set_ the progress bar to their desired production limit such that if they dragged it to 50% full example, that the machine would start and the progress bar would begin to _decrease_ until it reached 0. My reservation with this is that it would get complicated if they then wanted to set the production limit higher or especially lower than they originally intended. And because the fan rotation speed/progress bar does not expose details that would educate the user I decided to just keep it simple.

### Testing

BDD still, I am especially pleased to find that because of the polling interval, the production limit feature applies not only when the power is turned on from the apple home app, but also when the user instantiates ice production by pressing the power button on the machine. Basically anytime the production threshold is surpassed the machine shuts off.

Of note, I wasn't completely clear what OIM_PRODUCTION meant but went with the assumption of a percentage of ice production from 0 - 100. In reality the production probably gets up to about 120 before the machine actually shuts off. 
Via observation and a stopwatch, I was at least able to rule out that OIM_PRODUCTION meant number of minutes, because it takes longer than 140 minutes for the machine to fill.


### Reviewer Nudging

Take a look, and see if you agree with the production limit field living in the device section. It could alternatively live within advanced options, but in my opinion it should be particular to the device and not the plugin. Keep in mind that Advanced Option section is only exposed if one codes for a ProductionData characteristic as part of the AccessoryInformation service on a given device, its meant to be extendable basically.

I'd be eager to at least get these changes on a beta release tag for feedback purposes, but also fine to merge into latest as well.

jamesh48 committed:
initial icemaker production setup

jamesh48 committed:
production limit

jamesh48 committed:
support updateRate

jamesh48 committed:
dynamic values, fixes and cleanups

jamesh48 committed:
pre pr cleanups, comments

jamesh48 committed:
linting fixes

jamesh48 committed:
linting fixes

jamesh48 committed:
linting fixes

jamesh48 committed:
linting fixes

JasonGoldenDDT commented:
Awesome work!  Happy to test and share feedback.  

Use of RotationSpeed for tracking ice production is a win.  Despite the confusing translation.  It's a great start. 
Generally I never think of setting a production limit, I do want to see how much ice is available.  That is really helpful.  I live in a warm climate with out air conditioning, my home can reach 90 degrees during the day in the winter.  I know ice production is slower for me, the app always says 1 lb per hour... I'm curious to see how accurate that is now with this clock.

jamesh48 committed:
import Buffer

jamesh48 committed:
import Buffer

jamesh48 committed:
linting fixes

jamesh48 committed:
linting fixes

jamesh48 commented:
> Awesome work! Happy to test and share feedback.
> 
> Use of RotationSpeed for tracking ice production is a win. Despite the confusing translation. It's a great start. Generally I never think of setting a production limit, I do want to see how much ice is available. That is really helpful. I live in a warm climate with out air conditioning, my home can reach 90 degrees during the day in the winter. I know ice production is slower for me, the app always says 1 lb per hour... I'm curious to see how accurate that is now with this clock.

Thanks for the feedback!

I wonder if maybe the default value for production limit should be undefined and if so, evaluated internally as `Infinity` (which is a number in javascript), from my observation the machine still is making ice when production level reaches 100 and i'd hate to shortchange power users like yourself :)

jamesh48 committed:
linting fixes

jamesh48 committed:
linting fixes

jamesh48 commented:
@donavanbecker let me investigate my last comment before merging, regarding allowing the machine to complete its cycle if the production limit is unset.

@JasonGoldenDDT with this in mind, I'm thinking that in lieu of knowing what OIM_PRODUCTION actually means that we can test it _as_ a completion percentage and especially to see if 100% matches as reasonably close to completion for you in regards to the progress bar _view_, and I will still make it so that it at least keeps producing ice _after_ reaching 100 in the case of the production limit being unset.

donavanbecker commented:
I have it ready to merge into a beta branch so it will publish a beta once merge. So just let me know when you are ready.

jamesh48 committed:
PR Feedback

Update so Production Limit can be unset, allowing the opal ice machine to finish its production cycle naturally

jamesh48 committed:
lint fix

jamesh48 commented:
@donavanbecker its ready to go for a beta release

@JasonGoldenDDT 
In the homebridge settings, i'm sure you know that theres a switch toggle for turning on debug mode. Since its a release I made the logs regarding the ice production progress as debugSuccess. I also made it so that _only after_ the production value _exceeds_ 100 that this log begins to show a "Completion" value which continues to be the _actual_ production value that comes from the machine, the progress bar value itself maxes out at 100. 

It would be informative to compare notes on what that final completion value is when the machine finishes producing ice naturally. Maybe if its the same we can standardize on the upper limit.

In this example I just hardcoded 120 to illustrate what the log looks like.

<img width="558" alt="Screenshot 2025-02-26 at 12 51 41 PM" src="https://github.com/user-attachments/assets/6a6eef4d-988e-4ed8-819e-e4941f5fab24" />

JasonGoldenDDT commented:
Installed and testing set a 120 timer, will see where we end up.  

From the UI,  fan service is present and at Zero. 

feedback: I'm leaving the update intervals as they are for testing, but I would probably default to asking for updates every 3-5 minutes given how slow the ice production process is.   Generally the device is active for hours in our use case.. then off for a few hours before I turn it back on again for the even.  I would saw two people who only drink ice water can go through 2-3 bins in a 24 hour period?  my estimate.  Will narrow that down with this plugin. 

Updated - I hear the first ice drop at Value 9, pace picks up at 11.  I would classify 0-11 as "startup" and 11-100 as production.  It might make sense to start counting after 11 on a volume production meter. 

from the debug Log:
```
[2/26/2025, 10:15:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:16:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:16:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:16:58 AM] [SmartHQ] [DEBUG] data: {"item":{"applianceId":"2D9","erd":"0x9100","time":"2025-02-26T20:16:57.316Z","value":"05"},"resource":"/appliance/2D9/erd/0x9100","kind":"publish#erd","userId":"3z8"}
[2/26/2025, 10:16:58 AM] [SmartHQ] [DEBUG] ERD_CODES: OIM_STATUS
[2/26/2025, 10:16:58 AM] [SmartHQ] [DEBUG] obj>item>value: 05
[2/26/2025, 10:17:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:17:07 AM] [SmartHQ] [DEBUG] Opal Progress Svc: Value: 0, Limit: Infinity
[2/26/2025, 10:17:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:18:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:18:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:18:49 AM] [SmartHQ] [DEBUG] data: {"item":{"applianceId":"D9","erd":"0x9100","time":"2025-02-26T20:18:47.199Z","value":"06"},"resource":"/appliance/C2D9/erd/0x9100","kind":"publish#erd","userId":"3z8"}
[2/26/2025, 10:18:49 AM] [SmartHQ] [DEBUG] ERD_CODES: OIM_STATUS
[2/26/2025, 10:18:49 AM] [SmartHQ] [DEBUG] obj>item>value: 06
[2/26/2025, 10:19:06 AM] [SmartHQ] [DEBUG] data: {"item":{"applianceId":"2D9","erd":"0x9107","time":"2025-02-26T20:19:05.095Z","value":"01"},"resource":"/appliance/2D9/erd/0x9107","kind":"publish#erd","userId":"3z8"}
[2/26/2025, 10:19:06 AM] [SmartHQ] [DEBUG] ERD_CODES: OIM_POWER
[2/26/2025, 10:19:06 AM] [SmartHQ] [DEBUG] obj>item>value: 01
[2/26/2025, 10:19:06 AM] [SmartHQ] [DEBUG] data: {"item":{"applianceId":"02000034C2D9","erd":"0x9101","time":"2025-02-26T20:19:05.095Z","value":"01"},"resource":"/appliance/2D9/erd/0x9101","kind":"publish#erd","userId":"z8"}
[2/26/2025, 10:19:06 AM] [SmartHQ] [DEBUG] ERD_CODES: OIM_LIGHT_LEVEL
[2/26/2025, 10:19:06 AM] [SmartHQ] [DEBUG] obj>item>value: 01
[2/26/2025, 10:19:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:19:07 AM] [SmartHQ] [DEBUG] Opal Progress Svc: Value: 0, Limit: Infinity
[2/26/2025, 10:19:10 AM] [SmartHQ] [DEBUG] data: {"item":{"applianceId":"2D9","erd":"0x9100","time":"2025-02-26T20:19:09.101Z","value":"00"},"resource":"/appliance/2D9/erd/0x9100","kind":"publish#erd","userId":"s3z8"}
[2/26/2025, 10:19:10 AM] [SmartHQ] [DEBUG] ERD_CODES: OIM_STATUS
[2/26/2025, 10:19:10 AM] [SmartHQ] [DEBUG] obj>item>value: 00

// by default these "keep alive" seem to frequent. - 1-3 minutes default?
[2/26/2025, 10:19:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:20:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:20:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:21:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:21:08 AM] [SmartHQ] [DEBUG] Opal Progress Svc: Value: 0, Limit: Infinity
[2/26/2025, 10:21:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:22:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:22:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:23:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
[2/26/2025, 10:23:07 AM] [SmartHQ] [DEBUG] Opal Progress Svc: Value: 0, Limit: Infinity
[2/26/2025, 10:23:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
```

jamesh48 commented:
> Installed and testing set a 120 timer, will see where we end up.
> 
> From the UI, fan service is present and at Zero.
> 
> feedback: I'm leaving the update intervals as they are for testing, but I would probably default to asking for updates every 3-5 minutes given how slow the ice production process is. Generally the device is active for hours in our use case.. then off for a few hours before I turn it back on again for the even. I would saw two people who only drink ice water can go through 2-3 bins in a 24 hour period? my estimate. Will narrow that down with this plugin.
> 
> Updated - I hear the first ice drop at Value 9, pace picks up at 11. I would classify 0-11 as "startup" and 11-100 as production. It might make sense to start counting after 11 on a volume production meter.
> 
> from the debug Log:
> 
> ```
> [2/26/2025, 10:15:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:16:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:16:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:16:58 AM] [SmartHQ] [DEBUG] data: {"item":{"applianceId":"2D9","erd":"0x9100","time":"2025-02-26T20:16:57.316Z","value":"05"},"resource":"/appliance/2D9/erd/0x9100","kind":"publish#erd","userId":"3z8"}
> [2/26/2025, 10:16:58 AM] [SmartHQ] [DEBUG] ERD_CODES: OIM_STATUS
> [2/26/2025, 10:16:58 AM] [SmartHQ] [DEBUG] obj>item>value: 05
> [2/26/2025, 10:17:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:17:07 AM] [SmartHQ] [DEBUG] Opal Progress Svc: Value: 0, Limit: Infinity
> [2/26/2025, 10:17:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:18:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:18:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:18:49 AM] [SmartHQ] [DEBUG] data: {"item":{"applianceId":"D9","erd":"0x9100","time":"2025-02-26T20:18:47.199Z","value":"06"},"resource":"/appliance/C2D9/erd/0x9100","kind":"publish#erd","userId":"3z8"}
> [2/26/2025, 10:18:49 AM] [SmartHQ] [DEBUG] ERD_CODES: OIM_STATUS
> [2/26/2025, 10:18:49 AM] [SmartHQ] [DEBUG] obj>item>value: 06
> [2/26/2025, 10:19:06 AM] [SmartHQ] [DEBUG] data: {"item":{"applianceId":"2D9","erd":"0x9107","time":"2025-02-26T20:19:05.095Z","value":"01"},"resource":"/appliance/2D9/erd/0x9107","kind":"publish#erd","userId":"3z8"}
> [2/26/2025, 10:19:06 AM] [SmartHQ] [DEBUG] ERD_CODES: OIM_POWER
> [2/26/2025, 10:19:06 AM] [SmartHQ] [DEBUG] obj>item>value: 01
> [2/26/2025, 10:19:06 AM] [SmartHQ] [DEBUG] data: {"item":{"applianceId":"02000034C2D9","erd":"0x9101","time":"2025-02-26T20:19:05.095Z","value":"01"},"resource":"/appliance/2D9/erd/0x9101","kind":"publish#erd","userId":"z8"}
> [2/26/2025, 10:19:06 AM] [SmartHQ] [DEBUG] ERD_CODES: OIM_LIGHT_LEVEL
> [2/26/2025, 10:19:06 AM] [SmartHQ] [DEBUG] obj>item>value: 01
> [2/26/2025, 10:19:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:19:07 AM] [SmartHQ] [DEBUG] Opal Progress Svc: Value: 0, Limit: Infinity
> [2/26/2025, 10:19:10 AM] [SmartHQ] [DEBUG] data: {"item":{"applianceId":"2D9","erd":"0x9100","time":"2025-02-26T20:19:09.101Z","value":"00"},"resource":"/appliance/2D9/erd/0x9100","kind":"publish#erd","userId":"s3z8"}
> [2/26/2025, 10:19:10 AM] [SmartHQ] [DEBUG] ERD_CODES: OIM_STATUS
> [2/26/2025, 10:19:10 AM] [SmartHQ] [DEBUG] obj>item>value: 00
> 
> // by default these "keep alive" seem to frequent. - 1-3 minutes default?
> [2/26/2025, 10:19:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:20:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:20:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:21:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:21:08 AM] [SmartHQ] [DEBUG] Opal Progress Svc: Value: 0, Limit: Infinity
> [2/26/2025, 10:21:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:22:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:22:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:23:07 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> [2/26/2025, 10:23:07 AM] [SmartHQ] [DEBUG] Opal Progress Svc: Value: 0, Limit: Infinity
> [2/26/2025, 10:23:37 AM] [SmartHQ] [DEBUG] data: {"kind":"websocket#pong","id":"keepalive-ping"}
> ```

Yes that makes sense, I think from a development and testing perspective the updateRate of 30 seconds is more satisfying to see it work, but I can increase the default value there to 3-5 minutes (180-300) seconds. 

Regarding your follow up, for me it does not start counting essentially until ice is being produced, that your experience is different is also valid and we can consider the best approach.

I'm not sure about the websocket pings, thats just debug logs from the existing code that are coming along with the progress info that we are looking at.

Ok well thinking about it more, technically the websocket pings are there to maintain a persistent connection with the ice machine, thats why it says "keep alive". If there was an even more discreet level of debug logging that would be a better place for them, but I think the fact that they both are there and frequent informs us that the connection is stable, which is occasionally useful in the realm of debugging. The way I understand it is that within the realm of websocket protocol, that there is a timeout or heartbeat interval that has to be maintained between the two machines or the connection is closed. The communication model here is not like REST because the underlying connection stays open and the ping is just maintaining that, if we found a way to increase the time between pings, we may exceed the limit expected by the ice machine and then have an inconsistent connection that would continually need to reconnect.
